### PR TITLE
Move attribute naming docs from specification

### PR DESCRIPTION
**Pull request by jack-berg**
Related to https://github.com/open-telemetry/opentelemetry-specification/issues/3746. 

Moves [opentelemetry-specification/specification/common/attribute-naming.md](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/common/attribute-naming.md) and [opentelemetry-specification/specification/common/attribute-requirement-level.md](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/common/attribute-requirement-level.md) to `semantic-conventions`.

I've attempted to bring the commit history for these files along, but could use an extra set of eyes from @jsuereth, who I believe brought he git history with when seeding this repository.

### CODE REVIEW

**Comment by mx-psi** [comment]
I believe this fixes #295 cc @joaopgrassi

File: docs/general/attribute-naming.md [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+
+- When an attribute represents a single entity, the attribute name SHOULD be singular.
+  Examples: `host.name`, `db.user`, `container.id`.
+
+- When attribute can represent multiple entities, the attribute name SHOULD be pluralized
+  and the value type SHOULD be an array. E.g. `process.command_args` might include multiple
+  values: the executable name and command arguments.
+
+- When an attribute represents a measurement,
+  [Metric Name Pluralization Guidelines](https://github.com/open-telemetry/semantic-conventions/blob/main/docs/general/metrics.md#pluralization)
```
Comment (joaopgrassi) [inline docs/general/attribute-naming.md:68]: We can reference now the markdown file firectly

File: docs/general/attribute-requirement-level.md [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+- [Opt-In](#opt-in)
+- [Performance suggestions](#performance-suggestions)
+
+<!-- tocstop -->
+
+</details>
+
+_This section applies to Log, Metric, Resource, and Span, and describes requirement levels for attributes defined in semantic conventions._
+
+Attribute requirement levels apply to the [instrumentation library](../glossary.md#instrumentation-library).
```
Comment (joaopgrassi) [inline docs/general/attribute-requirement-level.md:22]: Now we will need to use the full link to the spec as we do elsewhere in the semconv repo. (referring to the spec version)

**Review by joaopgrassi — COMMENTED** [review]
Thank you!

The links needs to be adapted now, to link to the local files in this repo and use the full link to the spec repo.

File: docs/general/attribute-requirement-level.md [hunk 3]
```diff
@@ -0,0 +1,76 @@
+# Attribute Requirement Levels for Semantic Conventions
```
Comment (trask) [inline docs/general/attribute-requirement-level.md:1]:
```suggestion
# Attribute Requirement Levels
```

File: docs/general/attribute-naming.md [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+sure the name does not appear as a key of any "rename_attributes" section (keys
+denote old attribute names in rename operations).
+
+## Recommendations for OpenTelemetry Authors
+
+- All names that are part of OpenTelemetry semantic conventions SHOULD be part
+  of a namespace.
+
+- When coming up with a new semantic convention make sure to check existing
+  namespaces ([Semantic Conventions](https://github.com/open-telemetry/semantic-conventions/blob/main/docs/README.md))
```
Comment (trask) [inline docs/general/attribute-naming.md:86]:
```suggestion
  namespaces ([Semantic Conventions](../README.md))
```

File: docs/general/attribute-naming.md [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+  (more precisely to
+  [U+0021 .. U+007E](https://en.wikipedia.org/wiki/Basic_Latin_(Unicode_block)#Table_of_characters)
+  subset of Unicode code points). It is recommended to further limit names to
+  the following Unicode code points: Latin alphabet, Numeric, Underscore, Dot
+  (as namespace delimiter).
+
+## Recommendations for Application Developers
+
+As an application developer when you need to record an attribute first consult
+existing [semantic conventions](https://github.com/open-telemetry/semantic-conventions/blob/main/docs/README.md).
```
Comment (trask) [inline docs/general/attribute-naming.md:110]:
```suggestion
existing [semantic conventions](../README.md).
```

File: docs/general/attribute-requirement-level.md [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+- [Opt-In](#opt-in)
+
+The requirement level for an attribute is specified by semantic conventions depending on attribute availability across instrumented entities, performance, security, and other factors. When specifying requirement levels, a semantic convention MUST take into account signal-specific requirements.
+
+For example, Metric attributes that may have high cardinality can only be defined with `Opt-In` level.
+
+A semantic convention that refers to an attribute from another semantic convention MAY modify the requirement level within its own scope. Otherwise, requirement level from the referred semantic convention applies.
+
+<!-- TODO(jsuereth) - make examples not break on changes to semconv -->
+For example, [Database semantic convention](https://github.com/open-telemetry/semantic-conventions/blob/main/docs/database/README.md) references `network.transport` attribute defined in [General attributes](https://github.com/open-telemetry/semantic-conventions/blob/main/docs/general/README.md) with `Conditionally Required` level on it.
```
Comment (trask) [inline docs/general/attribute-requirement-level.md:38]:
```suggestion
For example, [Database semantic convention](../database/README.md) references `network.transport` attribute defined in [General attributes](./README.md) with `Conditionally Required` level on it.
```

File: docs/general/attribute-requirement-level.md [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+
+`http.route` is an example of a conditionally required attribute that is populated when the instrumented HTTP framework provides route information for the instrumented request. Some low-level HTTP server implementations do not support routing and corresponding instrumentations can't populate the attribute.
+
+When a `Conditionally Required` attribute's condition is not satisfied, and there is no requirement to populate the attribute, semantic conventions MAY provide special instructions on how to handle it. If no instructions are given and if instrumentation can populate the attribute, instrumentation SHOULD use the `Opt-In` requirement level on the attribute.
+
+<!-- TODO(jsuereth) - make examples not break on changes to semconv -->
+For example, `server.address` is `Conditionally Required` by the [Database convention](https://github.com/open-telemetry/semantic-conventions/blob/main/docs/database/README.md) when available. When `network.peer.address` is available instead, instrumentation can do a DNS lookup, cache and populate `server.address`, but only if the user explicitly enables the instrumentation to do so, considering the performance issues that DNS lookups introduce.
```
Comment (trask) [inline docs/general/attribute-requirement-level.md:55]:
```suggestion
For example, `server.address` is `Conditionally Required` by the [Database convention](../database/README.md) when available. When `network.peer.address` is available instead, instrumentation can do a DNS lookup, cache and populate `server.address`, but only if the user explicitly enables the instrumentation to do so, considering the performance issues that DNS lookups introduce.
```

**Review by mx-psi — APPROVED** [review]

File: docs/general/attribute-requirement-level.md [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+- [Opt-In](#opt-in)
+- [Performance suggestions](#performance-suggestions)
+
+<!-- tocstop -->
+
+</details>
+
+_This section applies to Log, Metric, Resource, and Span, and describes requirement levels for attributes defined in semantic conventions._
+
+Attribute requirement levels apply to the [instrumentation library](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/glossary.md#instrumentation-library).
```
Comment (joaopgrassi) [inline docs/general/attribute-requirement-level.md:22]:
```suggestion
Attribute requirement levels apply to the [instrumentation library](https://github.com/open-telemetry/opentelemetry-specification/blob/v1.26.0/specification/glossary.md#instrumentation-library)).
```

**Review by joaopgrassi — APPROVED** [review]

**Comment by jack-berg** [comment]
FYI @open-telemetry/specs-semconv-maintainers, I believe that when it comes time to merge this, we should preserve the commit history instead of squashing.

**Review by AlexanderWert — APPROVED** [review]

File: docs/general/attribute-requirement-level.md [hunk 9]
```diff
@@ -0,0 +1,78 @@
+# Attribute Requirement Levels
```
Comment (jsuereth) [inline docs/general/attribute-requirement-level.md:1]: We'll need to update all our internal links to the SPECIFICATION to this local document before you can delete the specification side, FYI.

That can be in this PR or a follow up.

**Review by pyohannes — APPROVED** [review]

File: docs/general/attribute-requirement-level.md (hunk 9 above)
Comment (jack-berg) [reply docs/general/attribute-requirement-level.md:1]: I've pushed a commit which updates all the references to the spec version of these documents I could find 👍

**Comment by arminru** [comment]
@chalin FYI - see also https://github.com/open-telemetry/opentelemetry-specification/pull/3758 🙂

**Review by arminru — APPROVED** [review]

**Review by chalin — APPROVED** [review]
LGTM - it doesn't break anything when published to the OTel website ✨ 

One nit, see inline comment.

File: docs/general/attribute-requirement-level.md
[the remark this answers is not in the document]
Comment (jack-berg) [reply docs/general/attribute-requirement-level.md:1]: Per convo with @chalin, I've formatted these files using:
```
npm run _check:format:any -- --write docs/general/attribute-*.md --prose-wrap=always
```

**Comment by jack-berg** [comment]
Reminder to whoever merges this:

> FYI @open-telemetry/specs-semconv-maintainers, I believe that when it comes time to merge this, we should preserve the commit history instead of squashing.

**Comment by AlexanderWert** [comment]
> Reminder to whoever merges this:
> 
> > FYI @open-telemetry/specs-semconv-maintainers, I believe that when it comes time to merge this, we should preserve the commit history instead of squashing.

@arminru @jsuereth `Rebase and merge` is not enabled for this repository:
image

Can we enable it temporarily? Or how can we achieve the above otherwise?

**Comment by chalin** [comment]
The commit history remains available in the PR, why impose that on the main branch? For projects that are likely to have a long life it is preferable, IMHO, to keep the main branch's history as clean and terse as possible. (My 2 cents :)

**Comment by jsuereth** [comment]
> The commit history remains available in the PR, why impose that on the main branch? For projects that are likely to have a long life it is preferable, IMHO, to keep the main branch's history as clean and terse as possible. (My 2 cents :)

@chalin The goal here is to preserve the previous history from the spec repo for the files that are moved into this repository.

@AlexanderWert I can temporarily enable for this PR.

**Comment by chalin** [comment]
> @chalin The goal here is to preserve the previous history from the spec repo for the files that are moved into this repository.

Oh, got it! 👍🏻

**Review by jsuereth — APPROVED** [review]

**Comment by jsuereth** [comment]
> This branch cannot be rebased due to conflicts

@jack-berg  do you mind manually rebasing this and `push -f` on your private branch?

**Comment by jack-berg** [comment]
>> This branch cannot be rebased due to conflicts

> @jack-berg do you mind manually rebasing this and push -f on your private branch?

Done 👍

**Pull request merged**